xray/src/inpaint.rs

```diff
@@ -0,0 +1,296 @@
+            for j in 0..height {
+                interpolate(i, j, right_weight, &mut right_sub, &mut current_sub);
+            }
+        }
+        current.save(current_path)?;
+        right.save(right_path)?;
+    }
+    Ok(())
+}
+
+fn interpolate_inpaint_image_with_bottom(
+    spatial_node_id: SpatialNodeId,
+    output_directory: &Path,
+) -> ImageResult<()> {
+    if let (Some((mut current, current_path)), Some((mut bottom, bottom_path))) = (
+        inpaint_image_and_path_from(spatial_node_id, output_directory, None)?,
+        inpaint_image_and_path_from(spatial_node_id, output_directory, Direction::Bottom)?,
+    ) {
+        let (width, height) = (current.width(), current.height() / 2);
+        let mut current_sub = current.sub_image(0, height, width, height);
+        let mut bottom_sub = bottom.sub_image(0, 0, width, height);
+        for j in 0..height {
+            let bottom_weight = j as f32 / (height - 1) as f32;
+            for i in 0..width {
+                interpolate(i, j, bottom_weight, &mut bottom_sub, &mut current_sub);
+            }
+        }
+        current.save(current_path)?;
+        bottom.save(bottom_path)?;
+    }
+    Ok(())
+}
+
+fn apply_inpainting(spatial_node_id: SpatialNodeId, output_directory: &Path) -> ImageResult<()> {
+    if let Some((inpaint_image, inpaint_image_path)) =
+        inpaint_image_and_path_from(spatial_node_id, output_directory, None)?
+    {
+        let (width, height) = inpaint_image.dimensions();
+        let image_view = inpaint_image.view(width / 4, height / 4, width / 2, height / 2);
+        let image_path = get_image_path(output_directory, NodeId::from(spatial_node_id));
+        image_view.to_image().save(image_path)?;
+        fs::remove_file(inpaint_image_path)?;
+    }
+    Ok(())
+}
+
+fn inpainting_step<P, F>(
+    message: &str,
+    pool: &Pool,
+    spatial_node_ids: &[SpatialNodeId],
+    partitioning_function: P,
+    spatial_node_function: F,
+) where
+    P: FnMut(&&SpatialNodeId) -> bool,
+    F: Fn(SpatialNodeId) -> ImageResult<()> + Send + Copy,
+{
+    let progress_bar = create_syncable_progress_bar(spatial_node_ids.len(), message);
+    let run_partition = |spatial_node_ids: Vec<SpatialNodeId>| {
+        pool.scoped(|scope| {
+            for spatial_node_id in spatial_node_ids {
+                let progress_bar = Arc::clone(&progress_bar);
+                scope.execute(move || {
+                    // TODO(feuerste): Move to rayon and try_for_each to handle errors properly!
+                    spatial_node_function(spatial_node_id).unwrap();
+                    progress_bar.lock().unwrap().inc();
+                });
+            }
+        });
+    };
+
+    let (first, second): (Vec<SpatialNodeId>, Vec<SpatialNodeId>) =
+        spatial_node_ids.iter().partition(partitioning_function);
+    run_partition(first);
+    run_partition(second);
+
+    progress_bar.lock().unwrap().finish_println("");
+}
+
+pub fn perform_inpainting(
```

Comment:
IMHO, it might make sense to have a structure like
```
struct Inpainter {
    pool: &Pool,
    output_directory: &Path,
    inpaint_distance_px: u8,
    leaf_node_ids: &FnvHashSet<NodeId>,
}
```
with the following methods:
`inpainting_step, perform_inpainting, create_inpaint_image,  interpolate_inpaint_image_with_right, interpolate_inpaint_image_with_bottom, apply_inpainting`, and perhaps other.
I think it would make it a little easier to use/read as you wouldn't have to pass pool, directory, and leaves to every function.

Reply:
I restructured the code to use two helper structs: One for the methods applying to a single node, one for the step methods. Please take another look!

Reply:
Looks good!